--- Xaw3d-1.5E/xc/lib/Xaw3d/Scrollbar.c.orig	Fri Apr 18 09:42:17 2003
+++ Xaw3d-1.5E/xc/lib/Xaw3d/Scrollbar.c	Fri Apr 18 09:44:12 2003
@@ -1087,7 +1087,11 @@
     Cardinal *num_params;	/* unused */
 {
     register ScrollbarWidget sbw = (ScrollbarWidget) w;
-    float top = sbw->scrollbar.top;
+    union {
+       XtPointer pt;
+       float     top;
+    } foo;
+    foo.top = sbw->scrollbar.top;
 
 #ifndef XAW_ARROW_SCROLLBARS
     if (sbw->scrollbar.direction == 0) return; /* if no StartScroll */
@@ -1116,10 +1120,10 @@
     /* Removed the dependancy on scrollbar arrows. Xterm as distributed in
        X11R6.6 by The XFree86 Project wants this correction, with or without
        the arrows. */
-    top += 0.0001;
+    foo.top += 0.0001;
 /* #endif */
-    XtCallCallbacks (w, XtNthumbProc, *(XtPointer*)&top);
-    XtCallCallbacks (w, XtNjumpProc, (XtPointer)&top);
+    XtCallCallbacks (w, XtNthumbProc, foo.pt);
+    XtCallCallbacks (w, XtNjumpProc, (XtPointer)&sbw->scrollbar.top);
 }
 
 
