--- Xaw3d-1.5/xc/lib/Xaw3d/List.c~	Tue Oct 15 16:41:20 1996
+++ Xaw3d-1.5/xc/lib/Xaw3d/List.c	Mon Jul 24 16:55:29 2000
@@ -37,6 +37,7 @@
  */
 
 #include <stdio.h>
+#include <stdlib.h>
 #include <ctype.h>
 
 #include <X11/IntrinsicP.h>
--- Xaw3d-1.5/xc/lib/Xaw3d/MultiSink.c~	Tue Oct 15 16:41:20 1996
+++ Xaw3d-1.5/xc/lib/Xaw3d/MultiSink.c	Mon Jul 24 16:58:09 2000
@@ -81,6 +81,7 @@
 #include <X11/Xaw3d/TextP.h>
 #include "XawI18n.h"
 #include <stdio.h>
+#include <stdlib.h>
 #include <ctype.h>
 
 #ifdef GETLASTPOS
--- Xaw3d-1.5/xc/lib/Xaw3d/Text.c~	Thu May 14 21:03:24 1998
+++ Xaw3d-1.5/xc/lib/Xaw3d/Text.c	Mon Jul 24 17:00:54 2000
@@ -71,6 +71,7 @@
 
 #include <X11/Xfuncs.h>
 #include <ctype.h>		/* for isprint() */
+#include <stdlib.h>		/* for abs() */
 
 #ifndef MAX_LEN_CT
 #define MAX_LEN_CT 6		/* for sequence: ESC $ ( A \xx \xx */


